File: src/main/java/core/procedures/interop/CharType.java
package core.procedures.interop;

import core.exceptions.WrongTypeException;
import core.procedures.AFn;
import core.procedures.FnArgsBuilder;

public final class CharType extends AFn {

  public CharType() {
    super(new FnArgsBuilder().min(1).max(1).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "char";
  }

  @Override
  public Character apply1(Object arg) {
    /* Have to box it */
    if (arg instanceof Number) {
      return (char)((Number) arg).intValue();
    } else if (arg instanceof Character) {
      return (Character) arg;
    }
    throw new WrongTypeException("char", "Character or Number", arg);
  }
}

File: src/main/java/core/procedures/interop/CharType.kt
package core.procedures.interop

import core.exceptions.WrongTypeException
import core.procedures.AFn
import core.procedures.Arity.Exactly

class CharType : AFn<Any?, Char>(name = "char", isPure = true, arity = Exactly(1)) {

    override operator fun invoke(arg: Any?) = when (arg) {
        is Number -> arg.toInt().toChar()
        is Char   -> arg
        else      -> throw WrongTypeException(name, "Character or Number", arg)
    }
}
